packages/ifpack2/src/Ifpack2_BlockTriDiContainer_impl.hpp

```diff
@@ -2442,17 +2422,22 @@ namespace Ifpack2 {
 	Kokkos::parallel_for
 	  (Kokkos::ThreadVectorRange(member, vector_loop_size),[&](const int &v) {
 	    solveMultiVector(member, blocksize, i0, r0, nrows, v, WW);
+	    copyToFlatMultiVector(member, partidx, npacks, pri0, v, blocksize, num_vectors);
 	  });
       }      
 
-      void run() {
+      void run(const impl_scalar_type_2d_view &Y,
+	       const impl_scalar_type_2d_view &Z) {
 #if defined(KOKKOS_ENABLE_CUDA) && defined(IFPACK2_BLOCKTRIDICONTAINER_ENABLE_PROFILE)
```

Comment:
It would be nice if we could combine all these different profiling and timing interface calls into a single function that doesn't have macros wrapped all around it.  How about this:
```c++
namespace Ifpack2 {
namespace Details {
  void startProfiling (const char label[]);
  void stopProfiling ();

  class ProfilingScope {
  public:
    ProfilingScope (const char label[]) { startProfiling (label); }
    ~ProfilingScope () { stopProfiling (); }
  };
}
}
```
The implementation of `startProfiling` could look like this:
```
void startProfiling (const char label[]) {
#ifdef WHATEVER_CUDA_STUFF
  cudaProfilerStart ();
#endif 
#ifdef WHATEVER_WE_LOVE_MACROS_I_GUESS
  TEUCHOS_FUNC_TIME_MONITOR(label);
#endif 
}
```

Reply:
Good idea. I will fix this way.

Reply:
(FYI the idea is that the implementation of these functions would go in a .cpp file, so that we wouldn't have to recompile Ifpack2's solvers if we change what Ifpack2 does for profiling.)

Reply:
BTW, teuchos time monitor cannot be inside of the function. The macro put an object to get the scope of timer. If we put the timer with the function, it only measure time for the startProfiling function.

Reply:
@kyungjoo-kim We could use the `Teuchos::Time` object directly:
```c++
namespace Ifpack2 {
namespace Details {

class ProfilingScope {
public:
  ProfilingScope ();
  ~ProfilingScope ();
private:
#ifdef WHATEVER_WE_LOVE_MACROS_I_GUESS
  Teuchos::RCP<Teuchos::Time> timer_;
#endif 
};

void ProfilingScope::startProfiling () {
  // ...
#ifdef WHATEVER_WE_LOVE_MACROS_I_GUESS
  // If the timer exists, return it, else make a new one.
  timer_ = Teuchos::TimeMonitor::getNewCounter (label);
  timer->start ();
#endif 
}

void ProfilingScope::startProfiling () {
  // ...
#ifdef WHATEVER_WE_LOVE_MACROS_I_GUESS
  timer_->stop ();
  timer_ = Teuchos::null;
#endif 
}
```

Reply:
when we define timer in this way, Is this also included in the summarize report ? and stacked timer ?

Reply:
@kyungjoo-kim The macro also calls `Teuchos::TimeMonitor::getNewCounter` (look in `teuchos/comm/src/Teuchos_TimeMonitor.hpp` for the macro definition).  If that works with stacked timer, then the above code should be fine.

Reply:
This teuchos timer does not accumulate time over multiple iterations.